contracts/WrapperRegistry.sol

```diff
@@ -0,0 +1,118 @@
+     *  address into `_underlyingToWrapperMapping`
+     */
+    function addWrapper(address wrapperToken) external override onlyOwner returns (bool) {
+        address underlyingToken = IButtonWrapper(wrapperToken).underlying();
+        if (_underlyingTokens.add(underlyingToken)) {
+            _underlyingToWrapperMapping[underlyingToken] = wrapperToken;
+            emit WrapperAdded(underlyingToken, wrapperToken);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * @inheritdoc IWrapperRegistry
+     * @dev Gets address of underlyingToken from `wrapperToken` and calls `removeUnderlying`
+     */
+    function removeWrapper(address wrapperToken) external override onlyOwner returns (bool) {
+        return removeUnderlying(IButtonWrapper(wrapperToken).underlying());
+    }
+
+    /**
+     * @inheritdoc IWrapperRegistry
+     * @dev Removes address of underlyingToken from `_underlyingTokens`
+     * @dev Removes the entry of underlyingToken address to wrapperToken
+     *  address from `_underlyingToWrapperMapping`
+     */
+    function removeUnderlying(address underlyingToken) public override onlyOwner returns (bool) {
+        if (_underlyingTokens.remove(underlyingToken)) {
+            emit WrapperRemoved(underlyingToken, _underlyingToWrapperMapping[underlyingToken]);
+            delete _underlyingToWrapperMapping[underlyingToken];
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * @inheritdoc IWrapperRegistry
+     * @dev Retrieves the length of `_underlyingTokens`
+     */
+    function numWrappers() public view override returns (uint256) {
+        return _underlyingTokens.length();
+    }
+
+    /**
+     * @inheritdoc IWrapperRegistry
+     * @dev No guarantees are made on the ordering.
+     * @dev Retrieves the underlying token at `index` and returns it with matching wrapper token
+     */
+    function wrapperAt(uint256 index) public view override returns (address, address) {
+        address underlyingToken = _underlyingTokens.at(index);
+        return (underlyingToken, _underlyingToWrapperMapping[underlyingToken]);
+    }
+
+    /**
+     * @inheritdoc IWrapperRegistry
+     * @dev Retrieves the wrapper from `_underlyingToWrapperMapping` by using
+     *  `underlyingToken` as the key
+     */
+    function getWrapperFromUnderlying(address underlyingToken)
+        external
+        view
+        override
+        returns (address)
+    {
+        return _underlyingToWrapperMapping[underlyingToken];
+    }
+
+    /**
+     * @inheritdoc IWrapperRegistry
+     * @dev Retrieves the underlying token from wrapperToken and returns it if present in registry
+     */
+    function getUnderlyingFromWrapper(address wrapperToken)
+        external
+        view
+        override
+        returns (address)
+    {
+        address underlyingToken = IButtonWrapper(wrapperToken).underlying();
+        if (
```

Comment:
@aalavandhan This is a non-permissioned view function that operates as the opposite of `getWrapperFromUnderlying`. The issue of not doing the check is that you can just take any wrapper and call `IButtonWrapper(wrapperToken).underlying()` without confirming if the wrapper is even in the registry or not. So if AMPL has one official wrapper and one unofficial bootleg wrapper, they'll both return AMPL even though only the first one is actually in the registry.

This isn't really a problem, but I think the point is that this function also operates as a check of "is this wrapper in the registry?". Otherwise people would just call `.underlying()` directly on the wrapper.